client/client.go

```diff
@@ -0,0 +1,768 @@
+	c.mu.Lock()
+	defer c.mu.Unlock()
+
+	c.userResponseHooks = append(c.userResponseHooks, h...)
+	return c
+}
+
+// JSONMarshal returns json marshal function in Core.
+func (c *Client) JSONMarshal() utils.JSONMarshal {
+	return c.jsonMarshal
+}
+
+// Set json encoder.
+func (c *Client) SetJSONMarshal(f utils.JSONMarshal) *Client {
+	c.jsonMarshal = f
+	return c
+}
+
+// JSONUnmarshal returns json unmarshal function in Core.
+func (c *Client) JSONUnmarshal() utils.JSONUnmarshal {
+	return c.jsonUnmarshal
+}
+
+// Set json decoder.
+func (c *Client) SetJSONUnmarshal(f utils.JSONUnmarshal) *Client {
+	c.jsonUnmarshal = f
+	return c
+}
+
+// XMLMarshal returns xml marshal function in Core.
+func (c *Client) XMLMarshal() utils.XMLMarshal {
+	return c.xmlMarshal
+}
+
+// Set xml encoder.
+func (c *Client) SetXMLMarshal(f utils.XMLMarshal) *Client {
+	c.xmlMarshal = f
+	return c
+}
+
+// XMLUnmarshal returns xml unmarshal function in Core.
+func (c *Client) XMLUnmarshal() utils.XMLUnmarshal {
+	return c.xmlUnmarshal
+}
+
+// Set xml decoder.
+func (c *Client) SetXMLUnmarshal(f utils.XMLUnmarshal) *Client {
+	c.xmlUnmarshal = f
+	return c
+}
+
+// TLSConfig returns tlsConfig in client.
+// If client don't have tlsConfig, this function will init it.
+func (c *Client) TLSConfig() *tls.Config {
+	if c.tlsConfig == nil {
+		c.tlsConfig = &tls.Config{
+			MinVersion: tls.VersionTLS12,
+		}
+	}
+
+	return c.tlsConfig
+}
+
+// SetTLSConfig sets tlsConfig in client.
+func (c *Client) SetTLSConfig(config *tls.Config) *Client {
+	c.tlsConfig = config
+	return c
+}
+
+// SetCertificates method sets client certificates into client.
+func (c *Client) SetCertificates(certs ...tls.Certificate) *Client {
+	config := c.TLSConfig()
+	config.Certificates = append(config.Certificates, certs...)
+	return c
+}
+
+// SetRootCertificate adds one or more root certificates into client.
+func (c *Client) SetRootCertificate(path string) *Client {
+	cleanPath := filepath.Clean(path)
+	file, err := os.Open(cleanPath)
+	if err != nil {
+		return c
+	}
+	defer func() {
+		_ = file.Close() //nolint:errcheck // It is fine to ignore the error here
+	}()
+	pem, err := io.ReadAll(file)
+	if err != nil {
+		return c
+	}
+
+	config := c.TLSConfig()
+	if config.RootCAs == nil {
+		config.RootCAs = x509.NewCertPool()
+	}
+	config.RootCAs.AppendCertsFromPEM(pem)
+
+	return c
+}
+
+// SetRootCertificateFromString method adds one or more root certificates into client.
+func (c *Client) SetRootCertificateFromString(pem string) *Client {
+	config := c.TLSConfig()
+
+	if config.RootCAs == nil {
+		config.RootCAs = x509.NewCertPool()
+	}
+	config.RootCAs.AppendCertsFromPEM([]byte(pem))
+
+	return c
+}
+
+// SetProxyURL sets proxy url in client. It will apply via core to hostclient.
+func (c *Client) SetProxyURL(proxyURL string) *Client {
+	pURL, err := urlPkg.Parse(proxyURL)
+	if err != nil {
+		log.Errorf("%v", err)
+
+		return c
+	}
+	c.proxyURL = pURL.String()
```

Comment:
Before this line we should add something like this to validate the proxies are http/https.

```go
    // Validate the URL scheme
    if pURL.Scheme != "http" && pURL.Scheme != "https" {
        return errors.New("unsupported proxy scheme")
    }
```